Create an Inventory of Known Technical Debt and Legacy Code. (#36)
Babylonpartners/ios-playbook: TechnicalDocuments/TechnicalDebt.md

## Objective-C code and GatewayManager.

GatewayManager supplies global dependencies to legacy Objective-C code. There was plenty of code in version one and two of the Babylon app that relied heavily on singletons.

| Objective-C view Controllers | Location | Comments |
| ---------------------------- | ---------| -------- |
| BillingInformationViewController | Clinical Records | |
| EditMembershipCodeViewController | Clinical Records | |
| MembershipTableViewController | Me Tab | |
| SurveyViewController | Clinical Records | |
| InsuranceViewController | Clinical Records | |
| BBAddAdditionalPatientInformationViewControllerV2 | Clinical Records | Might be dead code |

## Free Standing Swift View Controllers that should be Moved to Bento.

| View Controller | Comments |
| --------------- | -------- |

> | InfoViewController | Could be replaced with a Bento Component for scrolling through a couple of info screens. Only used for the GP @ hand on-boarding so could be moved from BabylonUI. |

Reviewer: You should add `IntroViewController` here. It's very similar to `InfoViewController` (indeed, I think there was talk of merging them at one point), but they some important differences in functionality. 

`IntroViewController` is used for quite a few flows (ID Verification, for example).

Author: 👍

| SegmentedViewController | Currently only used in Health Check to show the the two bands with organs and layers. Should probably be moved from BabylonUI. |

## View Models and Controllers that use Forms V2.

Forms V2 is no longer maintained and we should refactor these view controllers to use Bento.
Porting these view controllers to Bento will also make it easier to apply an app wide style guide (design library).

| View Model | Comments |
| ---------- | -------- |
| AppointmentCancellationViewModel | |
| AppointmentConsultationNotesViewModel | |
| AppointmentDetailsViewModel | |
| AppointmentNotesViewModel | |
| AppointmentPrescriptionViewModel | CE-30 |
| AppointmentReferralViewModel | |
| AppointmentReferralsViewModel | |
| AppointmentReplayViewModel | |
| AppointmentListViewModel | |
| BookAppointmentViewModel | |
| PractitionerDetailsViewModel | |
| ChatHistoryViewModel | |
| MedicalHistoryViewModel | |
| PersonalDetailsViewModel | |
| InfoItemsViewModel | |
| GPDetailsViewModel | |
| AddAddressViewModel | |
| PrivacySettingsViewModel | |
| ChooseCountryViewModel | |
| ForgotPasswordViewModel | |
| NotificationsViewModel | |
| ProfileViewModel | |

> | SignInViewModel | |

Reviewer 2: You might want to add NHSRegistrationStep1 and 2.

Technical debt and legacy code has a sad tendency to build up in any software project. As the iOS project has grown both in terms of team members and supported apps we have found that managing technical has become more difficult. Legacy code should be ideally be refactored to clear technical debt when changes are needed. Existing technical debt that should be addressed when working on a new feature should be included in the planning. There is an obvious risk that legacy code is maintained instead of refactored if it is not clear what the technical debt currently is.

## Objective-C code and GatewayManager.

GatewayManager supplies global dependencies to legacy Objective-C code. There was plenty of code in version one and two of the Babylon app that relied heavily on singletons.

| Objective-C view Controllers | Location | Comments |
| ---------------------------- | ---------| -------- |
| BillingInformationViewController | Clinical Records | |
| EditMembershipCodeViewController | Clinical Records | |
| MembershipTableViewController | Me Tab | |
| SurveyViewController | Clinical Records | |
| InsuranceViewController | Clinical Records | |

> | BBAddAdditionalPatientInformationViewControllerV2 | Clinical Records | Might be dead code |

Reviewer 3: > Might be dead code

`BBAddAdditionalPatientInformationViewControllerV2` is now only used for date of birth input via `BaseModel`. Screens: `EditMembershipCodeViewController` and `SurveyViewController`. If those are still in use, then first step could be to refactor `BBAddAdditionalPatientInformationViewControllerV2` to be used just for date of birth input. Cases other than `DateOfBirth` can be removed along with related code (e.g. `HeightAndWeightConverter`).

Reviewer 3: Also related to this (couldn't find a fitting section):
There was a transition period when two versions of Additional Info Bento existed. `AdditionalInfoViewModel`, which is no longer in use, along with related types can be removed. After this `AdditionalInfoViewModelV2` can be renamed to `AdditionalInfoViewModel`.

Author: Thanks for this. I think we can replace all remaining Objective-C view controllers in one go. If that is not feasible this background information will be very helpful.

> ## Free Standing Swift View Controllers that should be Moved to Bento.

Reviewer 4: > should be Moved to Bento

not sure about that, I'd leave this to be decided for later what exactly to do with them

Reviewer 4: Or do you mean that we should use something else from Bento instead of these controllers?

Author: These are form style view controllers that can be easily rewritten to use Bento.

There exist some other view controllers that don't fit what Bento currently support.

| View Controller | Comments |
| --------------- | -------- |
| InfoViewController | Could be replaced with a Bento Component for scrolling through a couple of info screens. Only used for the GP @ hand on-boarding so could be moved from BabylonUI. |

> | SegmentedViewController | Currently only used in Health Check to show the the two bands with organs and layers. Should probably be moved from BabylonUI. |

Reviewer 5: ```suggestion
| SegmentedViewController | Currently only used in Healthcheck to show the the two bands with organs and layers. Should probably be moved from BabylonUI. |
```
🙈

| IntroViewController | Could potentially be merged with InfoViewController. |

## View Models and Controllers that use Forms V2.

Forms V2 is no longer maintained and we should refactor these view controllers to use Bento.
Porting these view controllers to Bento will also make it easier to apply an app wide style guide (design library).

| View Model | Comments |
| ---------- | -------- |
| AppointmentCancellationViewModel | |
| AppointmentConsultationNotesViewModel | |
| AppointmentDetailsViewModel | |
| AppointmentNotesViewModel | |
| AppointmentPrescriptionViewModel | CE-30 |
| AppointmentReferralViewModel | |
| AppointmentReferralsViewModel | |
| AppointmentReplayViewModel | |
| AppointmentListViewModel | |
| BookAppointmentViewModel | |
| PractitionerDetailsViewModel | |
| ChatHistoryViewModel | |
| MedicalHistoryViewModel | |
| PersonalDetailsViewModel | |

> | InfoItemsViewModel | |

Reviewer 4: we should add `InfoRenderer`, we have a version in Bento, `InfoRendererV2`

Author: 👍

| GPDetailsViewModel | |
| AddAddressViewModel | |
| PrivacySettingsViewModel | |
| ChooseCountryViewModel | |
| ForgotPasswordViewModel | |
| NotificationsViewModel | |
| ProfileViewModel | |
| SignInViewModel | |
| NHSRegistrationStep1ViewController | |
| NHSRegistrationStep2ViewController | |

## Accessing Business Controllers Directly instead of using the SDK.

> Some of these business controllers are not defined inside the SDK. Most of them will be moved to the SDK, but some of them will stay in the Babylon project. Once the first, complete version of the SDK has been published there should not be any business controllers left in BabylonCore. Business controllers that are defined outside the SDK should have a documenting comment that explain why they are not part of the SDK.

Reviewer 6: This should mention that we must use all our SDK's.

Author: I am filling in a lot more detail about this in a document that I have started to write today.

| PractitionerBusinessController | BabylonAppointmentsUI |
| PatientBusinessControllerProtocol | BabylonAppointmentsUI, BabylonClinicalRecordsUI, BabylonUI, Babylon |
| PrescriptionBusinessControllerProtocol | BabylonAppointmentsUI, Babylon |
| PharmaciesBusinessControllerProtocol | BabylonAppointmentsUI, BabylonClinicalRecordsUI |
| ImageBusinessControllerProtocol | BabylonAppointmentsUI, BabylonClinicalRecordsUI, BabylonMonitor, Babylon |
| BookAppointmentBusinessControllerProtocol | BabylonAppointmentsUI |
| PrivacyBusinessControllerProtocol | BabylonAppointmentsUI, BabylonChatBotUI, BabylonUI |
| PublicHealthcareIdentifierBusinessController | BabylonAppointmentsUI, BabylonClinicalRecordsUI, BabylonUI, Babylon |
| SearchAssistantBusinessController | BabylonChatBotUI |
| MedicalHistoryBusinessController | BabylonClinicalRecordsUI |
| GenderBusinessControllerProtocol | BabylonClinicalRecordsUI, BabylonUI, Babylon |
| GeolocationBusinessController | BabylonClinicalRecordsUI, BabylonUI |
| InfoItemsBusinessControllerProtocol | BabylonClinicalRecordsUI |
| MapsBusinessControllerProtocol | BabylonClinicalRecordsUI |
| GPDetailsBusinessController | BabylonClinicalRecordsUI |

> | HealthcheckReportBusinessControllerProtocol | BabylonMonitor, BabylonUI, Babylon |

Reviewer 5: This likely won't become part of the SDK, since all the contents used inside the report web page is already available through the SDK anyway.

| ReferAFriendBusinessControllerProtocol | BabylonUI, Babylon |
| RatingBusinessController | BabylonUI |
| FamilyBusinessControllerProtocol | BabylonUI |
| RedemptionBusinessController | Babylon |
| RegionBusinessControllerProtocol | Babylon |
| EligibilityCheckBusinessControllerProtocol | Babylon |
| NotificationPayloadBusinessControllerProtocol | Babylon |
| SignInPrivacyBusinessControllerProtocol | Babylon |
| PaymentsBusinessControllerProtocol | Babylon |
| PDSQueueBusinessControllerProtocol | Babylon |
| LocationEligibilityBusinessController | Babylon |
| PostCodeEligibilityBusinessControllerProtocol | Babylon |
| SignUpBusinessController | Babylon |
| SignUpPrivacyBusinessController | Babylon |
| ForgotPasswordBusinessControllerProtocol | Babylon |
| FamilyBusinessControllerProtocol | Babylon |
| OnfidoBusinessControllerProtocol | Babylon |
| PDSBusinessControllerProtocol | Babylon |
| NHSBusinessController | Babylon |
| CreditCardsBusinessControllerProtocol | Babylon |
| TestsAndKitsBusinessControllerProtocol | Babylon |
| RedemptionBusinessControllerProtocol | Babylon |
| ForgotPasswordBusinessController | Babylon|

## Main Tab Bar and Navigation.

> Although much improved since version two was released, there is still a fair amount of confusing legacy code left in the tab bar flow controller and associated builders. Now that all content has been refactored to be based on reactive functional programming, flow controllers and builders it is possible to align `ExchangeAbleTabBarBuilder` and `TabContentBuilder` entities to follow our general design pattern. For historical reasons some flow controllers acts have logic for building view controllers. The `BabylonTabBarViewModel` should be able to forward all routing events to the appropriate flow controller. Some networking logic should be removed from `BabylonTabBarViewModel`. These deviations from our standard architecture makes it difficult to understand the code and changes often result in long and confusing discussions.

Reviewer 4: ```suggestion
Although much improved since version two was released, there is still a fair amount of confusing legacy code left in the tab bar flow controller and associated builders. Now that all content has been refactored to be based on reactive functional programming, flow controllers and builders it is possible to align `ExchangeableTabBarBuilder` and `TabContentBuilder` entities to follow our general design pattern. For historical reasons some flow controllers acts have logic for building view controllers. The `BabylonTabBarViewModel` should be able to forward all routing events to the appropriate flow controller. Some networking logic should be removed from `BabylonTabBarViewModel`. These deviations from our standard architecture makes it difficult to understand the code and changes often result in long and confusing discussions.
```

## Adapt the Code to Employ Current.

> Entities that have been moved to Current are still accessed via `AppDependencies.` Our agreed approach was to inject entities from Current into builder initialisers as default arguments. Before we extend Current to include more app wide content we should refactor how what is already available is accessed.

Reviewer 4: ```suggestion
Entities that have been moved to Current are still accessed via `AppDependencies`. Our agreed approach was to inject entities from Current into builder initialisers as default arguments. Before we extend Current to include more app wide content we should refactor how what is already available is accessed.
```
